larky/src/main/java/com/verygood/security/larky/modules/vgs/metrics/MetricsModule.java

```diff
@@ -56,65 +60,118 @@ public MetricsModule() {
         name = "bin",
         named = true,
+        positional = false,
         doc = "Bank Identification Number",
         allowedTypes = {
-          @ParamType(type = Integer.class)
+          @ParamType(type = StarlarkInt.class),
+          @ParamType(type = String.class),
+          @ParamType(type = NoneType.class),
         }
       ),
       @Param(
         name = "currency",
         named = true,
+        positional = false,
         doc = "Currency",
-        defaultValue = "NOT_SPECIFIED",
         allowedTypes = {
-          @ParamType(type = Currency.class)
+          @ParamType(type = String.class),
+          @ParamType(type = NoneType.class),
         }
       ),
       @Param(
         name = "psp",
         named = true,
+        positional = false,
         doc = "Payment Service Provider",
-        defaultValue = "NOT_SPECIFIED",
         allowedTypes = {
-          @ParamType(type = PSP.class)
+          @ParamType(type = String.class),
+          @ParamType(type = NoneType.class),
         }
       ),
       @Param(
         name = "result",
         named = true,
+        positional = false,
         doc = "Transaction Result",
-        defaultValue = "NOT_SPECIFIED",
         allowedTypes = {
-          @ParamType(type = TransactionResult.class)
+          @ParamType(type = String.class),
+          @ParamType(type = NoneType.class),
         }
       ),
       @Param(
         name = "type",
         named = true,
+        positional = false,
         doc = "Transaction Type",
-        defaultValue = "NOT_SPECIFIED",
         allowedTypes = {
-          @ParamType(type = TransactionType.class)
+          @ParamType(type = String.class),
+          @ParamType(type = NoneType.class),
+        }
+      ),
+      @Param(
+        name = "dictionary",
+        named = true,
+        positional = false,
+        doc = "Dictionary",
+        defaultValue = "{}",
+        allowedTypes = {
+          @ParamType(type = Dict.class)
         }
-      )
+      ),
     }
   )
-  @Override
   public void track(
-    int amount,
-    int bin,
-    Currency currency,
-    PSP psp,
-    TransactionResult result,
-    TransactionType type
+    Object amount,
+    Object bin,
+    Object currency,
+    Object psp,
+    Object result,
+    Object type,
+    Dict<String, Object> dictionary
   ) throws EvalException {
-    metrics.track(amount, bin, currency, psp, result, type);
+    metrics.track(
+      getInt(amount),
+      getInt(bin),
+      getValue(Currency.class, currency, Currency.NOT_SPECIFIED, Currency.UNKNOWN),
+      getValue(PSP.class, psp, PSP.NOT_SPECIFIED, PSP.UNKNOWN),
+      getValue(TransactionResult.class, result, TransactionResult.NOT_SPECIFIED, TransactionResult.UNKNOWN),
+      getValue(TransactionType.class, type, TransactionType.NOT_SPECIFIED, TransactionType.UNKNOWN),
+      dictionary
+    );
+  }
+
+  private static <T extends Enum<T>> T getValue(Class<T> enumClass, Object value, T notSpecified, T unknown) {
```

Comment:
Surprised if we don't already have the equivalent functions defined somewhere

Reply:
this is the function
`Enum.valueOf(enumClass, (String) value)`
It throws `NullPointerException` and `IllegalArgumentException`. I just used it for our purpose.

Reply:
no validation in `starlark`, validation moved to `protobuf`